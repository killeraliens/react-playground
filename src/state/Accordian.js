import React from 'react';


class Accordian extends React.Component {
  static defaultProps = {
    sections: []
  };

  constructor(props) {
    super(props);
    this.state = {
      current: null
    }
  }

  // componentDidMount() {
  //   this.setState({ current: this.props.sections[0]})
  // }

  handleButtonClick = (e) => {
    console.log(e.target.innerText);
    console.log(this.state.current);
    if (this.state.current !== null && e.target.innerText === this.state.current.title) {
       this.setState({ current: null });
    } else {
      this.setState({ current: this.props.sections[e.target.id]});
    }
  }

  renderSection = (section) => {
    if (this.state.current !== null && this.state.current === section) {
      // console.log('current section', this.state.current.title);
      return this.state.current.section;
    } else {
      return null;
    }
  }

  render() {
    return (
      <div>
        <ul>
          {this.props.sections.map((section, i) => {
            return (<li key={i} >
              <button id={i} onClick={this.handleButtonClick}>{section.title}</button>
              <div>{this.renderSection(section)}</div>
            </li>)
          })}
        </ul>
      </div>
    )
  }

}

export default Accordian;
