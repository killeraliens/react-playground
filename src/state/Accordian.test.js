import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {mount, shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import Accordian from './Accordian';

const sectionsProp = [
      {title: 'title 1', section: 'section 1 & some content'},
      {title: 'title 2', section: 'section 2 & some content'},
      {title: 'title 3', section: 'section 3 & some content'},
      {title: 'title 4', section: 'section 4 & some content'}
];

describe('Accordian component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Accordian />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders empty list to UI without props passed', () => {
    expect(
      renderer.create(<Accordian />).toJSON()
    ).toMatchSnapshot();
  });

  it('renders list items to UI with props passed', () => {
    expect(
      renderer.create(<Accordian sections={sectionsProp}/>).toJSON()
    ).toMatchSnapshot();
  });

  it('li buttons open section on click', () => {
    const wrapper = shallow(<Accordian sections={sectionsProp} />);
    wrapper.find('button').at(0).simulate('click', { target: {'id': 0} });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('shows only one section at a time', () => {
    const wrapper = shallow(<Accordian sections={sectionsProp} />);
    wrapper.find('button').at(1).simulate('click', { target: {'id': 1} });
    wrapper.find('button').at(3).simulate('click', { target: {'id': 3} });
    expect(toJson(wrapper)).toMatchSnapshot();
  })

  it('li button second click hides section', () => {
    const wrapper = shallow(<Accordian sections={sectionsProp} />);
    wrapper.find('button').at(0).simulate('click', { target: {'id': 0} });
    setTimeout(() => {
      wrapper.find('button').at(0).simulate('click', { target: {'id': 0} })
      const sectionContentDivText = wrapper.find('li').at(0).find('div').at(0).text();
      expect(sectionContentDivText).toEqual("");
    }, 1000);

  })

})
