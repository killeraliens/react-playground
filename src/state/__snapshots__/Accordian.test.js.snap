// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordian component li buttons open section on click 1`] = `
<div>
  <ul>
    <li
      key="0"
    >
      <button
        id={0}
        onClick={[Function]}
      >
        title 1
      </button>
      <div>
        section 1 & some content
      </div>
    </li>
    <li
      key="1"
    >
      <button
        id={1}
        onClick={[Function]}
      >
        title 2
      </button>
      <div />
    </li>
    <li
      key="2"
    >
      <button
        id={2}
        onClick={[Function]}
      >
        title 3
      </button>
      <div />
    </li>
    <li
      key="3"
    >
      <button
        id={3}
        onClick={[Function]}
      >
        title 4
      </button>
      <div />
    </li>
  </ul>
</div>
`;

exports[`Accordian component renders empty list to UI without props passed 1`] = `
<div>
  <ul />
</div>
`;

exports[`Accordian component renders list items to UI with props passed 1`] = `
<div>
  <ul>
    <li>
      <button
        id={0}
        onClick={[Function]}
      >
        title 1
      </button>
      <div />
    </li>
    <li>
      <button
        id={1}
        onClick={[Function]}
      >
        title 2
      </button>
      <div />
    </li>
    <li>
      <button
        id={2}
        onClick={[Function]}
      >
        title 3
      </button>
      <div />
    </li>
    <li>
      <button
        id={3}
        onClick={[Function]}
      >
        title 4
      </button>
      <div />
    </li>
  </ul>
</div>
`;

exports[`Accordian component shows only one section at a time 1`] = `
<div>
  <ul>
    <li
      key="0"
    >
      <button
        id={0}
        onClick={[Function]}
      >
        title 1
      </button>
      <div />
    </li>
    <li
      key="1"
    >
      <button
        id={1}
        onClick={[Function]}
      >
        title 2
      </button>
      <div />
    </li>
    <li
      key="2"
    >
      <button
        id={2}
        onClick={[Function]}
      >
        title 3
      </button>
      <div />
    </li>
    <li
      key="3"
    >
      <button
        id={3}
        onClick={[Function]}
      >
        title 4
      </button>
      <div>
        section 4 & some content
      </div>
    </li>
  </ul>
</div>
`;
